pkgs/by-name/ne/nexusmods-app/package.nix

```diff
@@ -67,6 +75,26 @@ buildDotnetModule rec {
 
   executables = [ meta.mainProgram ];
 
+  # FIXME: should some of these go in dotnetTestFlags and/or dotnetFlags?
+  dotnetBuildFlags =
+    let
+      # From https://nexus-mods.github.io/NexusMods.App/developers/Contributing/#for-package-maintainers
+      # TODO: add DefineConstants support to buildDotnetModule
```

Comment:
I don't feel like we need to specially add support for this when we already have a general build flag input.

From my experience it's rare to see any programs using DefineConstant to configure how they're built, usually they use MSBuild properties which is the idiomatic way to enable and disable things in the .NET world.

Reply:
> usually they use MSBuild properties which is the idiomatic way to enable and disable things in the .NET world.

The upstream project is still in and "experimental" state, so perhaps they'd be willing to migrate to a more idiomatic method then.

I'm somewhat unfamiliar with dotnet, so this is useful to know.

Reply:
We're trying to avoid build constants in the actual App as much as possible.
They're only really being used when it's desirable for package maintainers to have slightly different default runtime behaviour. Everything else is configurable at runtime.

For instance, for AppImage & Package Manager builds we add an extra message telling users to update using the System, rather than via the App. For manual installs we don't show update dialogs on the other hand.

It is possible to assign a MSBuild property other than `DefineConstants` here, or something like `AdditionalDefineConstants` that would just append to the existing constants. However at the end of the day you'll always be making some minor change. Be it adding a few characters to the build command, dropping a file, or replacing an enum value in the source to set some default behaviour at compile time.

Reply:
I meant that, usually, what projects do is have some msbuild property that then defines the constant, something like this in the `.csproj`:

```xml
  <PropertyGroup Condition="$(UseSystemExtractor) = '1'">
    <DefineConstants>$(DefineConstants);NEXUSMODS_APP_USE_SYSTEM_EXTRACTOR</DefineConstants>
  </PropertyGroup>
```

and then when building you'd do `dotnet build -c Release -p:UseSystemExtractor=1`.

So I don't know if it's worth adding `DefineConstants` support in `buildDotnetModule`for something that only one app will use most likely

Reply:
Ah, I've done it myself this way before too. I'd be happy to provide a constant like this upstream if the others there wouldn't mind.

Reply:
I don't think there's a need to change this now unless if there's an actual need for it, as it's been done like this already it'll just generate CI churn and problems for other packagers.

I'm just speaking on behalf of the .NET team in NixOS with relation to adding a builtin option for doing this, there's nothing inherently wrong with the way this has been done upstream, it's just not conventional so I think there's other things we can direct our effort towards.

Reply:
> I don't think there's a need to change this now unless if there's an actual need for it, as it's been done like this already it'll just generate CI churn and problems for other packagers.

I think it's still early enough in nexusmods-app's life that it shouldn't really be an issue for other downstream packagers.

Some of the issues @Sewer56 encountered while investigating https://github.com/Nexus-Mods/NexusMods.App/pull/1919 sound like they stemmed from us explicitly passing `DefineConstants` on the CLI. Long-term it may be better to have custom property groups, if so.

It's certainly easier to pass many `"-p:{}"` flags instead of joining a list of constants together using `"%3B"` (escaped `";"`)...

But I don't want to over-blow the issue; we can work with whichever implementation upstream prefers :grinning: